File: provider/src/main/java/com/raygun/raygun4android/CrashReporting.java
package com.raygun.raygun4android;

import android.content.ComponentName;
import android.content.Intent;
import android.os.Build;

import com.google.gson.Gson;
import com.raygun.raygun4android.logging.RaygunLogger;
import com.raygun.raygun4android.messages.crashreporting.RaygunBreadcrumbMessage;
import com.raygun.raygun4android.messages.crashreporting.RaygunMessage;
import com.raygun.raygun4android.network.RaygunNetworkUtils;
import com.raygun.raygun4android.services.CrashReportingPostService;
import com.raygun.raygun4android.utils.RaygunFileFilter;
import com.raygun.raygun4android.utils.RaygunFileUtils;
import com.raygun.raygun4android.utils.RaygunUtils;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class CrashReporting {

    private static RaygunUncaughtExceptionHandler handler;
    private static CrashReportingOnBeforeSend onBeforeSend;
    private static List tags;
    private static Map customData;
    private static List<RaygunBreadcrumbMessage> breadcrumbs = new ArrayList<>();
    private static boolean shouldProcessBreadcrumbLocation = false;

    static RaygunUncaughtExceptionHandler getExceptionHandler() {
        return CrashReporting.handler;
    }

    static void setOnBeforeSend(CrashReportingOnBeforeSend onBeforeSend) {
        CrashReporting.onBeforeSend = onBeforeSend;
    }

    static List getTags() {
        return CrashReporting.tags;
    }

    static void setTags(List tags) {
        CrashReporting.tags = tags;
    }

    static Map getCustomData() {
        return CrashReporting.customData;
    }

    static void setCustomData(Map customData) {
        CrashReporting.customData = customData;
    }

    static void recordBreadcrumb(String message) {
        RaygunBreadcrumbMessage breadcrumb = new RaygunBreadcrumbMessage.Builder(message).build();
        recordBreadcrumb(breadcrumb);
    }

    static void recordBreadcrumb(RaygunBreadcrumbMessage breadcrumb) {
        breadcrumbs.add(processBreadcrumbLocation(breadcrumb, shouldProcessBreadcrumbLocation(),3));
    }

    static void clearBreadcrumbs() {
        breadcrumbs.clear();
    }

    private static RaygunBreadcrumbMessage processBreadcrumbLocation(RaygunBreadcrumbMessage breadcrumb, boolean shouldProcessBreadcrumbLocation, int stackFrame) {

        if(shouldProcessBreadcrumbLocation && breadcrumb.getClassName() == null) {
            StackTraceElement frame = Thread.currentThread().getStackTrace()[stackFrame];

            return new RaygunBreadcrumbMessage.Builder(breadcrumb.getMessage())
                .category(breadcrumb.getCategory())
                .customData(breadcrumb.getCustomData())
                .level(breadcrumb.getLevel())
                .className(frame.getClassName())
                .methodName(frame.getMethodName())
                .lineNumber(frame.getLineNumber())
                .build();
        }

        return breadcrumb;
    }

    private static boolean shouldProcessBreadcrumbLocation() {
        return CrashReporting.shouldProcessBreadcrumbLocation;
    }

    static void shouldProcessBreadcrumbLocation(boolean shouldProcessBreadcrumbLocation) {
        CrashReporting.shouldProcessBreadcrumbLocation = shouldProcessBreadcrumbLocation;
    }

    static void send(Throwable throwable, List tags) {
        send(throwable, tags, null);
    }

    static void send(Throwable throwable, List tags, Map customData) {
        if (RaygunClient.isCrashReportingEnabled()) {

            RaygunMessage msg = buildMessage(throwable);

            if (msg == null) {
                RaygunLogger.e("Failed to send RaygunMessage - due to invalid message being built");
                return;
            }

            msg.getDetails().setTags(RaygunUtils.mergeLists(getTags(), tags));
            msg.getDetails().setCustomData(RaygunUtils.mergeMaps(getCustomData(), customData));

            if (onBeforeSend != null) {
                msg = onBeforeSend.onBeforeSend(msg);
                if (msg == null) {
                    return;
                }
            }

            enqueueWorkForCrashReportingService(RaygunClient.getApiKey(), new Gson().toJson(msg));
            postCachedMessages();
        } else {
            RaygunLogger.w("Crash Reporting is not enabled, please enable to use the send() function");
        }
    }

    private static RaygunMessage buildMessage(Throwable throwable) {
        try {
            RaygunMessage msg = RaygunMessageBuilder.instance()
                .setEnvironmentDetails(RaygunClient.getApplicationContext())
                .setMachineName(Build.MODEL)
                .setExceptionDetails(throwable)
                .setClientDetails()
                .setAppContext(RaygunClient.getAppContextIdentifier())
                .setVersion(RaygunClient.getVersion())
                .setNetworkInfo(RaygunClient.getApplicationContext())
                .setBreadcrumbs(breadcrumbs)
                .build();

            if (RaygunClient.getVersion() != null) {
                msg.getDetails().setVersion(RaygunClient.getVersion());
            }

            if (RaygunClient.getUser() != null) {
                msg.getDetails().setUserInfo(RaygunClient.getUser());
            } else {
                msg.getDetails().setUserInfo();
            }

            return msg;
        } catch (Exception e) {
            RaygunLogger.e("Failed to build RaygunMessage - " + e);
        }
        return null;
    }

    static void attachExceptionHandler() {
        Thread.UncaughtExceptionHandler oldHandler = Thread.getDefaultUncaughtExceptionHandler();
        if (!(oldHandler instanceof RaygunUncaughtExceptionHandler)) {
            CrashReporting.handler = new RaygunUncaughtExceptionHandler(oldHandler);
            Thread.setDefaultUncaughtExceptionHandler(CrashReporting.handler);
        }
    }

    static void postCachedMessages() {
        if (RaygunNetworkUtils.hasInternetConnection(RaygunClient.getApplicationContext())) {
            File[] fileList = RaygunClient.getApplicationContext().getCacheDir().listFiles(new RaygunFileFilter());
            if (fileList != null) {
                for (File f : fileList) {
                    try {
                        if (RaygunFileUtils.getExtension(f.getName()).equalsIgnoreCase(RaygunSettings.DEFAULT_FILE_EXTENSION)) {
                            ObjectInputStream ois = null;
                            try {
                                ois = new ObjectInputStream(new FileInputStream(f));
                                SerializedMessage serializedMessage = (SerializedMessage) ois.readObject();
                                enqueueWorkForCrashReportingService(RaygunClient.getApiKey(), serializedMessage.message);
                                if (!f.delete()) {
                                    RaygunLogger.w("Couldn't delete cached report (" + f.getName() + ")");
                                }
                            } finally {
                                if (ois != null) {
                                    ois.close();
                                }
                            }
                        }
                    } catch (FileNotFoundException e) {
                        RaygunLogger.e("Error loading cached message from filesystem - " + e.getMessage());
                    } catch (IOException e) {
                        RaygunLogger.e("Error reading cached message from filesystem - " + e.getMessage());
                    } catch (ClassNotFoundException e) {
                        RaygunLogger.e("Error in handling cached message from filesystem - " + e.getMessage());
                    }
                }
            } else {
                RaygunLogger.e("Error in handling cached message from filesystem - could not get a list of files from cache dir");
            }

        }
    }

    private static void enqueueWorkForCrashReportingService(String apiKey, String jsonPayload) {
        Intent intent = new Intent(RaygunClient.getApplicationContext(), CrashReportingPostService.class);
        intent.setAction("com.raygun.raygun4android.intent.action.LAUNCH_CRASHREPORTING_POST_SERVICE");
        intent.setPackage("com.raygun.raygun4android");
        intent.setComponent(new ComponentName(RaygunClient.getApplicationContext(), CrashReportingPostService.class));

        intent.putExtra("msg", jsonPayload);
        intent.putExtra("apikey", apiKey);

        CrashReportingPostService.enqueueWork(RaygunClient.getApplicationContext(), intent);
    }

    static class RaygunUncaughtExceptionHandler implements Thread.UncaughtExceptionHandler {
        private final Thread.UncaughtExceptionHandler defaultHandler;

        RaygunUncaughtExceptionHandler(Thread.UncaughtExceptionHandler defaultHandler) {
            this.defaultHandler = defaultHandler;
        }

        @Override
        public void uncaughtException(Thread thread, Throwable throwable) {

            List tags = new ArrayList();
            tags.add(RaygunSettings.CRASH_REPORTING_UNHANDLED_EXCEPTION_TAG);

            CrashReporting.send(throwable, tags);

            RUM.sendRemainingActivity();

            defaultHandler.uncaughtException(thread, throwable);
        }
    }
}

File: provider/src/main/java/com/raygun/raygun4android/CrashReporting.kt
package com.raygun.raygun4android

import android.os.Build
import com.google.gson.Gson
import com.raygun.raygun4android.logging.RaygunLogger
import com.raygun.raygun4android.messages.crashreporting.RaygunBreadcrumbMessage
import com.raygun.raygun4android.messages.crashreporting.RaygunMessage
import com.raygun.raygun4android.rum.RUM
import com.raygun.raygun4android.utils.RaygunUtils
import com.raygun.raygun4android.workers.CrashReportCache
import com.raygun.raygun4android.workers.CrashReportingWorkerHelper
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.util.concurrent.CopyOnWriteArrayList

typealias Tags = List<String>

typealias CustomData = Map<String, Any?>

object CrashReporting {
    private var exceptionHandler: RaygunUncaughtExceptionHandler? = null
    private var onBeforeSend: CrashReportingOnBeforeSend? = null
    private val coroutineScope = CoroutineScope(Dispatchers.IO + CoroutineName("CrashReporting"))

    @JvmField var tags: Tags? = null

    @JvmField var customData: CustomData? = null

    private val breadcrumbs: CopyOnWriteArrayList<RaygunBreadcrumbMessage> = CopyOnWriteArrayList()
    private var shouldProcessBreadcrumbLocation = false

    @JvmStatic
    fun setOnBeforeSend(onBeforeSend: CrashReportingOnBeforeSend?) {
        CrashReporting.onBeforeSend = onBeforeSend
    }

    @JvmStatic
    fun recordBreadcrumb(message: String?) {
        val breadcrumb = RaygunBreadcrumbMessage.Builder(message).build()
        recordBreadcrumb(breadcrumb)
    }

    @JvmStatic
    fun recordBreadcrumb(breadcrumb: RaygunBreadcrumbMessage) {
        breadcrumbs.add(processBreadcrumbLocation(breadcrumb, shouldProcessBreadcrumbLocation()))
    }

    @JvmStatic
    fun clearBreadcrumbs() {
        breadcrumbs.clear()
    }

    private fun processBreadcrumbLocation(
        breadcrumb: RaygunBreadcrumbMessage,
        shouldProcessBreadcrumbLocation: Boolean,
    ): RaygunBreadcrumbMessage {
        if (shouldProcessBreadcrumbLocation && breadcrumb.className == null) {
            val trace = Thread.currentThread().stackTrace
            var frame: StackTraceElement? = null

            if (trace.isNotEmpty()) {
                for (i in 0..<trace.size - 1) {
                    val thisFrame = trace[i]
                    val nextFrame = trace[i + 1]

                    if (
                        thisFrame.className.contains("com.raygun.raygun4android.") &&
                        !nextFrame.className.contains("com.raygun.raygun4android.")
                    ) {
                        frame = nextFrame
                        break
                    }
                }
            }

            if (frame != null) {
                return RaygunBreadcrumbMessage
                    .Builder(breadcrumb.message)
                    .category(breadcrumb.category)
                    .customData(breadcrumb.customData)
                    .level(breadcrumb.level)
                    .className(frame.className)
                    .methodName(frame.methodName)
                    .lineNumber(frame.lineNumber)
                    .build()
            }
        }

        return breadcrumb
    }

    private fun shouldProcessBreadcrumbLocation(): Boolean = shouldProcessBreadcrumbLocation

    @JvmStatic
    fun shouldProcessBreadcrumbLocation(shouldProcessBreadcrumbLocation: Boolean) {
        CrashReporting.shouldProcessBreadcrumbLocation = shouldProcessBreadcrumbLocation
    }

    @JvmStatic
    @JvmOverloads
    fun send(
        throwable: Throwable,
        tags: Tags?,
        customData: CustomData? = null,
    ) {
        if (RaygunClient.isCrashReportingEnabled) {
            coroutineScope.launch {
                val jsonPayload = buildJsonPayload(throwable, tags, customData) ?: return@launch
                enqueueWorkForCrashReporting(RaygunClient.apiKey, jsonPayload)
                postCachedMessages()
            }
        } else {
            RaygunLogger.w(
                "Crash Reporting is not enabled, please enable to use the send() function",
            )
        }
    }

    private suspend fun buildJsonPayload(
        throwable: Throwable,
        tags: Tags?,
        customData: CustomData?,
    ): String? {
        var msg = buildMessage(throwable)

        if (msg == null) {
            RaygunLogger.e("Failed to send RaygunMessage - due to invalid message being built")
            return null
        }

        msg.details.tags = RaygunUtils.mergeLists(CrashReporting.tags, tags)
        msg.details.customData = RaygunUtils.mergeMaps(CrashReporting.customData, customData)

        if (onBeforeSend != null) {
            msg = onBeforeSend!!.onBeforeSend(msg) ?: return null
        }

        return Gson().toJson(msg)
    }

    private fun cacheUnhandledException(
        throwable: Throwable,
        tags: Tags,
    ) {
        runBlocking(Dispatchers.IO) {
            val jsonPayload = buildJsonPayload(throwable, tags, null) ?: return@runBlocking
            CrashReportCache.store(RaygunClient.getApplicationContext(), jsonPayload)
        }
    }

    private suspend fun buildMessage(throwable: Throwable): RaygunMessage? {
        try {
            val msg =
                RaygunMessageBuilder()
                    .setEnvironmentDetails(RaygunClient.getApplicationContext())
                    .setMachineName(Build.MODEL)
                    .setExceptionDetails(throwable)
                    .setClientDetails()
                    .setAppContext(RaygunClient.appContextIdentifier)
                    .setVersion(RaygunClient.version)
                    .setNetworkInfo(RaygunClient.getApplicationContext())
                    .setBreadcrumbs(breadcrumbs)
                    .build()

            if (RaygunClient.version != null) {
                msg.details.version = RaygunClient.version
            }

            if (RaygunClient.user != null) {
                msg.details.user = RaygunClient.user
            } else {
                msg.details.setUserInfo()
            }

            return msg
        } catch (e: Exception) {
            RaygunLogger.e("Failed to build RaygunMessage - $e")
        }
        return null
    }

    @JvmStatic
    fun attachExceptionHandler() {
        val oldHandler = Thread.getDefaultUncaughtExceptionHandler()
        if (oldHandler !is RaygunUncaughtExceptionHandler && oldHandler != null) {
            exceptionHandler = RaygunUncaughtExceptionHandler(oldHandler)
            Thread.setDefaultUncaughtExceptionHandler(exceptionHandler)
        }
    }

    @JvmStatic
    fun postCachedMessages() {
        coroutineScope.launch {
            for (file in CrashReportCache.files(RaygunClient.getApplicationContext())) {
                CrashReportingWorkerHelper.enqueueCachedCrashReport(
                    RaygunClient.getApplicationContext(),
                    file,
                    RaygunClient.apiKey,
                )
            }
        }
    }

    private fun enqueueWorkForCrashReporting(
        apiKey: String?,
        jsonPayload: String,
    ) {
        CrashReportingWorkerHelper.enqueueCrashReport(
            RaygunClient.getApplicationContext(),
            jsonPayload,
            apiKey,
        )
    }

    class RaygunUncaughtExceptionHandler(
        private val defaultHandler: Thread.UncaughtExceptionHandler,
    ) : Thread.UncaughtExceptionHandler {
        override fun uncaughtException(
            thread: Thread,
            throwable: Throwable,
        ) {
            val tags = listOf(RaygunSettings.CRASH_REPORTING_UNHANDLED_EXCEPTION_TAG)
            try {
                cacheUnhandledException(throwable, tags)
                RUM.instance.sendRemaining()
            } catch (exception: Exception) {
                RaygunLogger.e("Failed to cache unhandled exception: $exception")
            } finally {
                defaultHandler.uncaughtException(thread, throwable)
            }
        }
    }
}
